### PR TITLE
Cut corners

### PR DESCRIPTION
**Pull request by pi-squared-studio**
image

> [!CAUTION]
> After conducting several tests, I came to the conclusion that in this form, this PR is not effective enough. This PR is relevant only for models not exceeding 2-3 cm in size with a standard nozzle of 0.4. In addition, it turned out that in addition to improving, this PR introduces small other impairments that do not cover the positive effect.  All visible improvements made by this PR are private, and are eliminated using other methods:
> - The main problem is the protruding corners when the lines are superimposed. They do not always occur, but only when the exterior wall is printed first and at certain printing speeds. This is not a mathematical (geometric) defect, but rather akin to a mechanical ringing. It is easily eliminated by selecting the Junction Deviation, PA parameter, changing the wall printing order or decreasing printer speed.
> - Forming the inner and outer corners - you can simply reduce the printing speed or use a smaller diameter nozzle.
> - Precision of lines and smooth surfaces for artistic printing - reducing the flow by 3-5%.
>
> There were also positive aspects in the implementation of this PR, they can already be applied in completely different conceptual additions, these are:
> - Deliberately installing the nozzle in a corner for the predicted flow near. Some kind of analogue of canceling Junction Deviation or similar algorithms.
> - The correct formation of the ends of the printed track for the implementation of "beautiful" seams.
> - Formation of an acute corner more than 135 degrees.
>
> At the moment, the project remains open for about a week (starting on January 20) to complete the discussion and close some issues. Next, I'll close it.

On the left is the usually maded print, on the right with the Improved Corners function enabled. Printed on one bed side to side. You can see that the angle on the modified sample is completely even. But this example is not perfect, as there are also defects caused by the resonance of the extruder.  But this is already a significant result, because the basic negative factor for other settings has been eliminated. This resonance has also become easily detectable - you can see the downward waves at the bottom of the right sample, as they were masked by excessive extrusion during normal printing. This PR generally enhances the geometric characteristics of the print.

# The preamble
All the time I've been doing 3D printing, I've noticed that the corners оа prints always turn out to be convex. This is especially noticeable when using silk plastics. Previously, I always associated this effect with transients in the extruder. I also thought that various parameterization methods such as Jerk, Junction Deviation, Pressure Advance and Input Shaping... But, as it turned out, it was difficult to remove the protruding corners. In some ways, I was getting better results, but in another place there was a severe deterioration.
I began to analyze this problem, and came to the conclusion that it is purely mechanical, but  all the above methods of adjustment manipulated the speed. When I printed the same models at a very low speed, these defects did not go away, which means they did not depend in any way on the speed changes in the movement of the printhead and the extrusion speed.

# Theory
If we analyze the path of the nozzle movement near the corner, we can see that excessive plastic extrusion occurs at a short distance. It comes from the fact that the Slicer operates with a certain ideal algorithm, which in practice turns out to be not exactly like that. 

image image image image

In these drawings, areas with denser coloring show an overflow of material at the the lines junctions or in the their corners.

The line in the slicer is filled with some rectangular segments, which, when intersected, give a small area of a quarter of the nozzle area where 2 times more filament is squeezed out, the volume of which would be enough to fill an empty square. But since the nozzle has a circular cross-section, and the slicer operates with rectangular paths, more plastic is obviously squeezed out in the corners about by 21%. Moreover, the angle formed by the nozzle is not straight, but at least repeats the roundness of the nozzle.

That is, in real conditions, excess plastic will not flow towards the corner, but a kind of plaque will be formed that will distort the correct shape of the corner.

image

I tried several algorithms for forming an angle, including creating a flow gradient on the approach to this angle, but they did not produce the expected result. And reproducing the gradient at such distances turned out to be even more problematic, since it was necessary to dose volumes comparable to single steps of the extruder. I decided to make an algorithm that simply symmetrically shortens the extruded lines depending on the angle. But at the same time, the nozzle still makes its previous trajectory into the corner.

image image image

# Realization

image

In the process of long research, I came to the conclusion that the formation of the exact angle is influenced by 2 factors - the superposition of lines near the junctions, and the superposition of lines at the formed corner.  Separating these factors allowed me to write very simple and lightweight code without much mathematical calculations. 

##  Superposition of lines near the junctions
This factor is always applied when there is a connection between two lines. It is very difficult to calculate the required volume for extraction, knowing that many other transients may occur, so there is a changeable overlap parameter **cut_corners_overlap** for more precise adjustment.

image image image

Mathematical calculations indicate an optimal value of 0.79, but it can be reduced to 0.5 depending on the material properties. If you select a value of 1.0, the corner overlap will be the same as when cut_corners is disabled, but the flow compensation will remain when overlapping lines at sharp angles.
Default value is 0.785398 - the difference between the area of a real round and square nozzle hole. The last statement is used as the basis for calculating of filament consumption volume of any slicer.

> [!WARNING]
> Since the ends of the lines are cutted by this method, in the g-code viewer may contain an unusual picture. But it everything is as it should be. With this method, the lines are shortened to compensate for excess material at the ends, but for the shortest segments, the flow density decreases without affecting the print accuracy. The prints will turn out to be smooth in reality without these visible tears.
>
> image image
> Figure: The flow rate map with disabled (at left) and enabled (at right) option "Cut Corners"
> It is possible that the position of the seams will not be displayed. To configure them, disable this option for a while.

##  Superposition of lines at the formed corner

image
image image

At this stage, the refilling area is simply tracked through the tangent of the angle and the superimposed line is filled using a flow gradient.

image

# Usage

This option eliminates overflow in the connection of two printed lines when excessive material is extruded at their ends.
Positive moment is especially noticeable on the protruding corners of the model.
This also affects to decrease the overflow of a solid infill (like as Archimedean Chords) if it consists of many small segments.

> [!CAUTION]
> The actions of this option are basic for others additional flow- and speed- shapers such as Jerk, Junction Deviation, Pressure Advance, and Input Shaping... Therefore, when calibrating, first set the **cut_corners_overflow** parameter and then recalibrate the rest from this list. Also set new values of the overlapping of seams, infills and perimeters.
> The effect of this option does not depend on the print speed, but choose slow speeds for calibration to eliminate other factors that affect accuracy.
> The correct results will be on the side that is less affected by the resonance effect of the print head.

## Conclusion and plans
At the moment, some types of lines, such as sloping ones and arcs, fall under the influence of this PR, but this is only a small part of all prints.  Maybe I'll try to implement this in the future.

### CODE REVIEW

**Comment by ianalexis** [comment]
I've tried to implement something similar here #11403 reducing flow but desisted for lack of time.
If this works correctly will close https://github.com/OrcaSlicer/OrcaSlicer/issues/11254

Quick tip, i think that to get the correct value we should calibrate PA with the tower and checking only the seams. Then we can calibrate this based in how the corners looks.

File: src/libslic3r/PrintConfig.cpp [hunk 1]
```diff
@@ -4433,6 +4433,21 @@ void PrintConfigDef::init_fff_params()
     // start and end point is from the change_filament_gcode
     def->set_default_value(new ConfigOptionPoints{Vec2d(30, -3), Vec2d(54, 245)});
 
+    def          = this->add("improved_corners", coBool);
+    def->label   = L("Improve corners");
```
Comment (ianalexis) [inline src/libslic3r/PrintConfig.cpp]: I think that improve corners is too generic. Maybe something like "cut cornet" or "corner overlap compensation" will be better.

**Comment by pi-squared-studio** [comment]
> I've tried to implement something similar here #11403 reducing flow but desisted for lack of time.
> If this works correctly will close #11254

Oh, how much information I missed!  Honestly, I haven't seen it, although I'm trying to read everything.

**Comment by pi-squared-studio** [comment]
> Quick tip, i think that to get the correct value we should calibrate PA with the tower and checking only the seams. Then we can calibrate this based in how the corners looks.

At this stage, I have disabled all modifiers so that they do not affect the experiment. Because I believe that this problem is of a basic nature, and primarily affects the results being deleted in Pressure Advance.

**Comment by pi-squared-studio** [comment]
@ianalexis Yes, I also tried to control the flow at the corners, but I didn't get very good results, since the shape of the corner turned out to be convex anyway. I attribute this to the fact that there is a dynamic distribution of plastic in an unpredictable way between an already printed and an as-yet-unprinted line. The best result was shown by the option with symmetrical interruption of the flow.

File: src/libslic3r/PrintConfig.cpp (hunk 1 above)
Comment (pi-squared-studio) [reply src/libslic3r/PrintConfig.cpp]: Okay, then I'll translate this PR into a draft, and if there are any other comments, I'll fix everything later.

**Comment by pi-squared-studio** [comment]
OK, I've found a solution to this problem.
As it turned out, there are 2 factors involved in the formation of the corner bulb, which is that the mathematics of calculating the flow of extruded material using a parallelepiped (if the nozzle has а square hole) does not correspond to the volume at the ends of the extruded material using a round pass nozzle. At each junction point, even on a straight line, 0.032 mm3 more was squeezed out (with a layer height of 0.2 and a line width of 0.4).
The second factor is the overlap of the lines when forming the angle. This factor was also calculated and embedded in the code. It performs quite well at angles up to 135 degrees. There are plans to refine by gradient filling for the large angles.
This algorithm also had a good effect on the quality of the upper surfaces and the condition of the seams. Now, knowing about these factors, I have reason to believe that the flow rate calibration test according to Archimedean Chords gives wrong results, since the infill has many joints in itself.

I'll post photos with the tests and a description of this process a bit later.

**Comment by pi-squared-studio** [comment]
Here are the photos of the test. On the left is a regular print, on the right with the enabled function.
It's very difficult to show anything in the photo, but in reality it looks much more confident.
Don't pay attention to the dots on the surface, it's either a plastic problem or an extruder problem. there wasn't time to figure out the problem yet.
Also, I still have a visible protrusion formed by resonance on one side. But it is much smaller in shape than without the enabled function of this PR. For the test, I turned off all additional extrusion line shapers such as Jerk and Input Shaping. Junction Deviation remained at a minimum of 0.001 - at the moment it is difficult to disable it for this test.
image
image
image
image
image
image
image
image

**Comment by RF47** [comment]
@pi-squared-studio this PR can Close #11254

**Comment by pi-squared-studio** [comment]
> @pi-squared-studio this PR can Close #11254

I don't think so. This PR was developed from scratch and describes a completely different view of the ongoing processes.
In that process, there is a reduction in extrusion at the ends, in my case, excess plastic is being calculated and excluded from the printing process.
I'll bring it to perfection, and it's up to the users to decide what to do.

**Comment by pi-squared-studio** [comment]
image
The purity of the sharp angle (15 degrees).

**Comment by pi-squared-studio** [comment]
One wall test
image
image

File: src/libslic3r/GCode.cpp [hunk 2]
```diff
@@ -6578,33 +6578,41 @@ std::string GCode::_extrude(const ExtrusionPath &path, std::string description,
                     }
                     if (sloped == nullptr) {
                         // Normal extrusion
-                        if (m_config.cut_corners && path.role() != erGapFill) { // Orca: Сut corners
-                            Vec2d  _a = this->point_to_gcode(line.a);
-                            Vec2d  _c(this->point_to_gcode(line.b) - _a);
+                        Vec2d &_b = this->point_to_gcode(line.b); // end point
+                        if (m_config.cut_corners) { // Orca: Сut corners 
+                            Vec2d &_a = this->point_to_gcode(line.a); // start point
```
Comment (RF47) [inline src/libslic3r/GCode.cpp]: maybe this solve the issue
```cpp
                        Vec2d _b = this->point_to_gcode(line.b); // end point
                        if (m_config.cut_corners) { // Orca: Сut corners 
                            Vec2d _a = this->point_to_gcode(line.a); // start point
```

Comment (pi-squared-studio) [reply src/libslic3r/GCode.cpp]: Okay, I'll even make constants. )))

**Comment by pi-squared-studio** [comment]
I update the topic with description for Wiki

**Comment by pi-squared-studio** [comment]
I rewrote the flow calculation algorithm in corners a dozen times. At the moment it looks like this. All the points are taken into account with meticulous precision, such as the superposition of corners and their order, as well as the volume of plastic in the corners. I made an algorithmic model without interrupting the flow. In fact, it turned out to be a more accurate model of flow compensation for small areas.
image
The code is coming soon.

**Comment by RF47** [comment]
Left normal, right cut corners:
image
image
Can a angle threshold angle be added to ignore smooth curves?

**Comment by RF47** [comment]
I think arc fitting is not compatible with this feature:
image

**Comment by RF47** [comment]
Up: normal
down: cut corners
It's not very noticeable in the photo, but the corners are much more sharply defined with cut corners.
image

**Comment by pi-squared-studio** [comment]
Thanks for the tests. But I've reworked this code again. I reduced the nozzle resolution to 1/2 of the diameter.
The results are completely different there. The code is very branched due to various conditions and extremely small extrusion values.  Now I'll optimize everything in the code to the end and send it. 

In the old code, the line was interrupted at the nodes to compensate for excess plastic. The size of this gap varied depending on the direction of printing, the seam, or the angle at which everything was reduced. But such interruptions gave artifacts, I had to do the same thing, but by changing the flow (but not the speed).

**Comment by RF47** [comment]
> Thanks for the tests. But I've reworked this code again. I reduced the nozzle resolution to 1/2 of the diameter. The results are completely different there. The code is very branched due to various conditions and extremely small extrusion values. Now I'll optimize everything in the code to the end and send it.
> 
> In the old code, the line was interrupted at the nodes to compensate for excess plastic. The size of this gap varied depending on the direction of printing, the seam, or the angle at which everything was reduced. But such interruptions gave artifacts, I had to do the same thing, but by changing the flow (but not the speed).

I'll try it again when it's available. I think the quality has improved significantly with these enhancements.

**Comment by pi-squared-studio** [comment]
@RF47 corrected the code
Now everything looks like this (the arrows show the seam positions (this is a test for different positions of the superimposed lines). The corner is very tiny, about 2.5 by 6 mm.
image

In fact, I'm more interested in the purity and detail of artistic prints. That's how the flow redistribution on it.
image

**Comment by boromyr** [comment]
I downloaded the latest build and I get this strange behavior when I enable wipe before external perimeter, even when the corner gag is disabled.
image
while when I activate it
image

**Comment by pi-squared-studio** [comment]
@boromyr Please drop the flow rate map. And it looks very strange, because this PR should not affect the shape of the lines. It looks like there's a horizontal scarf option involved, which I don't know anything about.

**Comment by boromyr** [comment]
No, no scarf stitching, without wipe before external perimeter everything works correctly. What do you mean by "Please drop the flow rate map"?

**Comment by pi-squared-studio** [comment]
> No, no scarf stitching, without wipe before external perimeter everything works correctly. What do you mean by "Please drop the flow rate map"?

Yes, I just reproduced these conditions. This is a small rotated segment that is added to the perimeter. Everything looks correct to me. But I suspect that the algorithm logically reduces its volume without this, since there is always an excess of it at the ends. And it becomes zero and it piece drops out. The code for it will not generated.
Yes, maybe I'll fix this point.

image

**Comment by boromyr** [comment]
image

**Comment by pi-squared-studio** [comment]
@boromyr  I removed the command that skips moving the extruder when the volume is minimal. It should be working fine now. 
Perhaps this segment will not be printed, then just disable this cleaning for a while. I suspect that now the usual seams should be almost perfect, since excess plastic will no longer be physically present. Try to set the seam clearance to 0 or at least a minimum.

**Comment by boromyr** [comment]
I completed the last commit and now the preview is correct; I will try to do some test prints as soon as possible.  
I noticed that the flow rate smoothing works in synergy with corner cutting; is that correct or are they two incompatible functions?  
Why is it not found in the quality tab?

**Comment by pi-squared-studio** [comment]
> I completed the last commit and now the preview is correct; I will try to do some test prints as soon as possible. I noticed that the flow rate smoothing works in synergy with corner cutting; is that correct or are they two incompatible functions? Why is it not found in the quality tab?

Initially, Cut corners (CC) was a G-code modifier, so it was placed on this tab. But there will be no problems to transfer to any place on any tab. This is how it will be more convenient for the end user.

At the expense of synergy, this option is echoed only by Small area flow compensation (SAFC), and is more resource-intensive.
To avoid double processing when choosing CC, the SAFC algorithm bypasses only on critical printing elements such as perimeters and infill.
Please note that the Jerk or Junction Deviation settings should be in the program settings and the printer should be zero. Otherwise, they will smooth out all the beneficial effects of this program. The printing speed will naturally decrease. But that's the price for quality.
Another note is that when using CC, the Pressure Advance option is welcome, but Input Shaping can affect the quality of corner shaping if manual calibration has been performed.

**Comment by pi-squared-studio** [comment]
I discovered a funny fact - that it is necessary not to reduce the amount of plastic in the corners, but rather to add it, compensating for the fact that the actual length of the line is equal to the length of the slicer line + the diameter of the nozzle (this is without taking into account the fact that it needs to be slightly removed at the corners due to overlapping lines).
And the quality of the corners themselves is affected by the fact that there are additional break points of the line in the PR. Although the speed of movement remains constant. Apparently, the printer's algorithms perceive this as a small printable element and do not include it in the global calculations of the movements. I use a Marlin. Maybe things are different on the Klipper.
image image

**Comment by pi-squared-studio** [comment]
It turns out to be very complex mathematics, in a very simple algorithm. There is no need to resort to complex trigonometric arithmetic. The calculation is done by interpolating from the angle when superimposing lines, and adding and removing material in places where there is a lot or little of it. If you wish, I will try to describe the logical explanation of this theory. 

Yes, one more thing - now the ends of the polyline are formed precisely enough without shapeless bumps. This means that it is necessary to observe the gap of the usual seams about the distance of the nozzle diameter, or 100%. I won't say that they turned out to be perfect, but at least they were thin enough, smooth on the surface and neat.

I also request to check, now it is possible to add more plastic to the corners if algorithms for rounding are used like Jerk or Junction Deviation. Just set cut_corners_overlap > 1.0.

**Comment by pi-squared-studio** [comment]
I've simplified the algorithm a bit. The scheme of work is now like this. I remind you that it is now possible to compensate for the lack of plastic in the corners, in case some algorithms round them off.

image

**Comment by ianalexis** [comment]
> I remind you that it is now possible to compensate for the lack of plastic in the corners, in case some algorithms round them off.

I have to test it but if this work as it should i think that we need to calibrate the PA by looking the seam in the PA Tower with a Seam gap = 0.
Then we can add a calibration test for this using something like the SCV model and increase the cut corner overlap step by step.
What do you think?

**Comment by ianalexis** [comment]
Notes for the future,

- Before this is merged:
  - Maybe we should set this value to 0 in some calibration tests like PA (see plater.cpp)
  - Test the relation with PA
- After this is merged:
  - We can create a calibration for this value.

**Comment by pi-squared-studio** [comment]
> I have to test it but if this work as it should i think that we need to calibrate the PA by looking the seam in the PA Tower with a Seam gap = 0.

Small remarks - PA was turned off in this test.  But still, some improvement in the angles was noticed here.
image
In any case, the inclusion of PA will affect the filling of this transition seam. For the test, we will need to determine the methodology so that the impact of PA on the results is minimal. 

We can talk about extremely small changes that cannot be determined by the eye. Even when I was making comparative models, I only determined when I knew which result to look at. The average user may not notice the detectable results, or interpret them in a biased manner. 

Regarding the gap, an algorithm has now been applied that if there is an open loop, or a closed one with a gap greater than 3/4 of the nozzle diameter, then the end of the line with the required filling volume will be formed. Otherwise, there will be a smooth transition.

The algorithm also determines the angles of the line displacement, if they are less than 30 degrees, then the line is drawn smoothly without changing the printing speed. If it is larger, then at each such node the printhead will stop to form the shape of the corner. I suspect that in this case, such shapers as Jerk or Junction Deviation (I don't know how for Klipper)  they won't work. The overall printing speed will really drop, but the drawing of the corners will be excellent.

In connection with the above, I do not know which test to use to identify the benefits of this PR. I use c printing with other flow formers disabled.  And I trust the mathematical model. )))

**Comment by celtare21** [comment]
Don't know if this is intended, but I tested a small cylinder with the recommended values (0.6 cut corners overlap and 100% nozzle diamater seam gap).
This is with Cut Corners disabled:
{3BCAB37C-79EB-47B2-95F4-17C33841B24D}
This is with it enabled:
{FC812180-215A-422F-BC56-6BE38FD551AD}

It shows very many retract moves.

Edit:

This is visible in the final print:
PXL_20260117_170244664

**Comment by pi-squared-studio** [comment]
@celtare21 Thanks for the test. Now I'll see what might be wrong. It is possible that the retraction causes a negative flow, which is formed when calculating short lines. I've fix this case.
Regarding the remaining deformations, it is possible that factors such as PA and IS cause them, although I had turned them off for the test. Now I will look at how the algorithm of this PA interacts with them.  
For a more precise study, I recommend enabling Verbose g-Code. It will be clearer at what stage the problem occurs. Thank you.

**Comment by celtare21** [comment]
[Cylinder.zip](https://github.com/user-attachments/files/24691190/Cylinder.zip)
Here is the G-code with verbose enabled. It is pretty easy to replicate, it's the built in orca cylinder primitive scaled to 20mm, I ran it with 0% infill, 3 walls, 6 top 5 bottom

**Comment by ianalexis** [comment]
I see an issue with seams, its only showing one object seam line.
This PR


imagen

Last nigthly
imagen


[test cut corners.3mf.txt](https://github.com/user-attachments/files/24691503/test.cut.corners.3mf.txt)

**Comment by pi-squared-studio** [comment]
> [Cylinder.zip](https://github.com/user-attachments/files/24691190/Cylinder.zip) Here is the G-code with verbose enabled. It is pretty easy to replicate, it's the built in orca cylinder primitive scaled to 20mm, I ran it with 0% infill, 3 walls, 6 top 5 bottom

In the latest version, I fixed this error with retracts. It was due to the fact that at small angles and short distances about nozzle diameter, their flow value could go much less than zero, which was perceived as a retraction. The defects on your figure were most likely due to excessive retractions.
Now I'm going to try to print this model for the sake of interest, although it looks smooth without it.
image

**Comment by pi-squared-studio** [comment]
@ianalexis 
It is very difficult to evaluate the work of this PR on my printer. The fact is that there is a third-party effect that hides visible defects on the print without this PR. I can say that there is an excess of material ( talking about the volume of approximately 0,4*0,2*0,2 mm) is smeared across the line and causes some moire. When I remove this extra volume in this PR, the perimeters become smooth, but other natural defects open up.  

Here are examples: on the left - without the effect of this PR, on the right - with PR. The printing direction of the coils is from right to left. The left corner is control (in fact, the surface is smoother, I checked with my fingernail, and the contrasting track is due to the optical refraction of light at the corner). The inner groove on the right is a natural niche caused by a lack of material due to the convergence of the walls.

This is without PA
image

with PA (0.045 in my case)
image

It seems we need to come up with another test to effectively evaluate the work of this PR. I'm going to print out a simple cube.

**Comment by ianalexis** [comment]
Yesterday i did some testing of different values of cut corner and tbh didint see much (if any).
Let me know when you are comfortable with the implementation and will test it all again.

**Comment by pi-squared-studio** [comment]
> Yesterday i did some testing of different values of cut corner and tbh didint see much (if any). Let me know when you are comfortable with the implementation and will test it all again.

the last implementation is the most correct.
I printed out the cube. Yes, there are changes, but now there is a visible resonance from an out-of-tune IS, which I have not seen before, and therefore have never calibrated. In this light, you can see that the relief has become smoother, and the edges do not bend like a bowl (at least on the right side, where there is no resonance).

image
image
image

Well, here's what the resonance looks like, which is now visible due to the precise dosing of plastic during printing. It is not possible to see anything in the photo. Unless you look at the light from a small angle. Then with PR, it seems a little more detailed printing. Again, in many ways everything is subjective.

image

**Comment by pi-squared-studio** [comment]
@ianalexis  Now I will try not to add plastic to the inner corners. I'll send you an update soon.

**Comment by celtare21** [comment]
Looks much better but there seems like there's so sporadic unretract moves:
{F1A5DE9A-8D01-41AD-905D-54FBAB87968A}
{FACC9AC4-F443-4085-9A22-811BF5723C04}
{3E071516-BC81-4F1C-80A0-A187333D67D6}


Don't know if they're intentional

**Comment by pi-squared-studio** [comment]
@celtare21 

It is ok. The fact is that in such corners, compensation for a certain amount of plastic is required. Relatively speaking, by shortening the lines, we eliminate the dense zone when overlaying the lines, while the white zone will increase slightly. It is at this point that the necessary material is added to form the correct corner. In the viewer, this is defined as Unrectract.

image image image

image

**Comment by celtare21** [comment]
Question, is the recommended value for cut corners overlap still 0.6? With 100% seam gap?

**Comment by pi-squared-studio** [comment]
The last test:

image
image
image
image

Right is CC: PA=0, IS=OFF, cut_corners_overlap=1.2 (yes, now possible grow it for visually contrast)
When using CC, you need to disable PA or set the minimum, as it hides all the advantages of this PR by rounding off the corners.

The main addition is the trimming of the inner corners and a special wiping-like movement for more precise shaping both inners and outers ones. These are micro-movements that are performed to form a fine structure more clearly. At this point, the nozzle stops completely, so the overall printing speed will be lower.

**Comment by pi-squared-studio** [comment]
> Question, is the recommended value for cut corners overlap still 0.6? With 100% seam gap?

the overlap is calculated as follows:
0.6 - for precise dosing of the material to form an angle with a radius equal to 1/2 of the nozzle diameter
0.7 - from this point on, micro-movement will occur to the outside of the corner to compensate for rounding
1.0 - is the volume of plastic, if the corner is considered completely square
3.0 - 3 radii, the maximum amount of compensation for rounding

for the gap:
100% - with such a gap, the ends of the lines touch without visible swelling.
50% is the minimum value when the end of the line (heh, I called it bulb) will be formed.
< 50% - the line will be smooth (this condition is necessary for a smooth transition between the superimposed lines)
image

**Comment by pi-squared-studio** [comment]
> [!CAUTION]
> After conducting several tests, I came to the conclusion that in this form, this PR is not effective enough. This PR is relevant only for models not exceeding 2-3 cm in size with a standard nozzle of 0.4. In addition, it turned out that in addition to improving, this PR introduces small other impairments that do not cover the positive effect.  All visible improvements made by this PR are private, and are eliminated using other methods:
> - The main problem is the protruding corners when the lines are superimposed. They do not always occur, but only when the exterior wall is printed first and at certain printing speeds. This is not a mathematical (geometric) defect, but rather akin to a mechanical ringing. It is easily eliminated by selecting the Junction Deviation, PA parameter, changing the wall printing order or decreasing printer speed.
> - Forming the inner and outer corners - you can simply reduce the printing speed or use a smaller diameter nozzle.
> - Precision of lines and smooth surfaces for artistic printing - reducing the flow by 3-5%.
>
> There were also positive aspects in the implementation of this PR, they can already be applied in completely different conceptual additions, these are:
> - Deliberately installing the nozzle in a corner for the predicted flow near. Some kind of analogue of canceling Junction Deviation or similar algorithms.
> - The correct formation of the ends of the printed track for the implementation of "beautiful" seams.
> - Formation of an acute corner more than 135 degrees.
>
> At the moment, the project remains open for about a week (starting on January 20) to complete the discussion and close some issues. Next, I'll close it.
